import os, glob, sys
import subprocess as sp
from itertools import groupby

configfile: "../config/config.json"

# Get custom parameters
input_fasta = config.get("input_fasta")  # path
step_size = config.get("step_size")  # int
chunks = config.get("chunks")  # True or None
chunk_size = config.get("chunk_size")  # int or None
strand = config.get("strand")  # 'both', '+' or '-'
batch_size = config.get('batch_size')
output_name_ = config.get('output_name')
ext = config.get("output_type")

include: "rules/common.smk"
include: "rules/download.smk"

if not config.get("download_model"):
    if chunks == True:
        CHR_ = get_chr_names(input_fasta, chunks, chunk_size)
    else:
        CHR_dict = get_chr_names(input_fasta, chunks, chunk_size)
        CHR_ = list(CHR_dict.keys())
        config['CHR_dict'] = CHR_dict
    
    config['CHR_'] = CHR_
    rule all:
        """ Run SnoBIRD to predict C/D box snoRNA genes in a genomic sequence."""
        input:
            final_output = expand('results/final/{output_name}.{output_type}', output_type=ext, output_name=output_name_)

rule all_downloads:
    """ Download all models that constitute SnoBIRD."""
    input:
        model1 = "data/references/models/snoBIRD_first_model.pt",
        model2 = "data/references/models/snoBIRD_second_model.pt",
        tokenizer = "data/references/DNA_BERT_6_tokenizer/",
        dnabert = "data/references/DNA_BERT_6_pretrained_model/"



include: "rules/snoBIRD.smk"
#include: "rules/target_prediction.smk"

# Define which rule to run depending if only SnoBIRD's first model 
# is run or if both models are run 
if config.get("first_model_only") == True:
    ruleorder: find_sno_limits_shap_minimal > filter_sno_pseudo_predictions_with_features
else:
    ruleorder: filter_sno_pseudo_predictions_with_features > find_sno_limits_shap_minimal
